climada/util/coordinates.py

```diff
@@ -76,6 +78,273 @@
+        
+    See also
+    --------
+    * util.coordinates.compute_geodesic_lengths()
+    * util.lines_polys_handler.point_exposure_from_lines()
+    """
+
+    if not isinstance(gdf_lines, gpd.GeoDataFrame):
+        gdf_lines = gpd.read_file(gdf_lines)
+        
+    gdf_points = gdf_lines.copy()
+    gdf_points['length'] = np.ones(len(gdf_points))*point_dist
+    gdf_points['length_full'] = compute_geodesic_lengths(gdf_points)
+
+    # split line lengths into relative fractions acc to point_dist (e.g. 0, 0.5, 1)
+    gdf_points['distance_vector'] = gdf_points.apply(
+        lambda row: np.linspace(0, 1, num=int(np.ceil(row.length_full/
+                                                      row.length)+1)), axis=1)
+
+    # create MultiPoints along the line for every position in distance_vector
+    gdf_points['geometry'] = gdf_points.apply(
+        lambda row: MultiPoint(
+            [row.geometry.interpolate(dist, normalized=True) 
+             for dist in row.distance_vector]), axis=1)
+    
+    # expand gdf from MultiPoint entries to single Points per row
+    return gdf_points.explode().drop(['distance_vector', 'length_full'], axis=1)
+
+def interpolate_polygons(gdf_poly, area_point):
+    """For a GeoDataFrame with polygons, get equally distributed lat/lon pairs
+    throughout the geometries, at a user-specified area distance
+    
+    Parameters
+    ----------
+    gdf_poly : gpd.GeoDataFrame
+        with polygons to be interpolated
+    area_point : float
+        area in m2 which one point should represent
+    
+    Returns
+    -------
+    (gpd.GeoDataFrame) with individual Point per row, retaining all other column infos
+        belonging to its corresponding polygon
+    
+    See also
+    --------
+    * util.coordinates.compute_geodesic_lengths()
+    * util.lines_polys_handler.point_exposure_from_polygons()
+    """
+    
+    metre_dist = math.sqrt(area_point)
+    gdf_points = gdf_poly.copy()
+    
+    gdf_points['degree_dist'] = gdf_poly.apply(lambda row: metres_to_degrees(
+        row.geometry.representative_point().x, 
+        row.geometry.representative_point().y,
+        metre_dist), axis=1)
+    
+    # get params to make an even grid with desired resolution 
+    # over bounding boxes of polygons:
+    gdf_points[['height','width','trans']] = gdf_points.apply(
+        lambda row: pts_to_raster_meta(row.geometry.bounds, 
+                                       (row.degree_dist, -row.degree_dist)), 
+        axis=1).tolist()
+    
+    # make grid --> lon / lat for each polygon-bbox
+    gdf_points[['lon','lat']] = pd.DataFrame(gdf_points.apply(lambda row: 
+        raster_to_meshgrid(row.trans, row.width, row.height), axis=1).tolist())
+    
+    for axis in ['lon', 'lat']:
+        gdf_points[axis] = gdf_points.apply(lambda row: row[axis].flatten(), 
+                                            axis=1)
+    # filter only centroids in actual polygons
```

Comment:
If I am not mistaken, you can thus easily end up with a polygon that is mapped to no point at all (if it is between raster points). I am not sure how to solve this, but it probably is not desired behaviour, or at least it should be warned about.

One possibility is to assign a value proportional to the cell area in a given polygon. This will lead to points belonging to different polygons at the same time though.

Reply:
Well, if you choose your representative area per point too small, then yes. The function assumes that the user adequately judges the size of the polygons in relation to the area per point. Aka, if you have a tiny island of 1km^2, you shouldn't interpolate your polygon with a `point_area` of 10'000m2

Reply:
There's also a lot of discussions online on how to best allocate a certain amount of points within an arbitrary 2D shape. There's a few fancy algorithms out there, but honestly, i think for our purpose, the commonly accepted easy-way out is good enough. And that's making an even grid of pre-determined resolution, and then cropping it to the outline of the polygon. This is how it's implemented atm

Reply:
If you have say 100 polygons, and one happens to be very slim, then you get problems too, irrespective of the area per point. But I agree that this might be ok, but I would at least raise a warning if one of the polygons has no point associated.

Reply:
e.g. (https://stackoverflow.com/questions/11178414/algorithm-to-generate-equally-distributed-points-in-a-polygon)

Reply:
A warning is now raised if a polygon has no point with the given resolution. Also, one "representative" point is then still set into the shape.